src/order/interval.lean

```diff
@@ -0,0 +1,422 @@
+
+/-- The injection from which we get the order. -/
+def to_dual_prod : nonempty_interval α → αᵒᵈ × α := to_prod
+
+@[simp] lemma to_dual_prod_apply (a : nonempty_interval α) :
+  a.to_dual_prod = (to_dual a.fst, a.snd) := prod.mk.eta.symm
+
+lemma to_dual_prod_injective : injective (to_dual_prod : nonempty_interval α → αᵒᵈ × α) :=
+λ a b, (ext_iff _ _).2
+
+instance [is_empty α] : is_empty (nonempty_interval α) := ⟨λ a, is_empty_elim a.fst⟩
+instance [subsingleton α] : subsingleton (nonempty_interval α) :=
+to_dual_prod_injective.subsingleton
+
+instance : has_le (nonempty_interval α) := ⟨λ a b, b.fst ≤ a.fst ∧ a.snd ≤ b.snd⟩
+
+lemma le_def : a ≤ b ↔ b.fst ≤ a.fst ∧ a.snd ≤ b.snd := iff.rfl
+
+/-- Turn an interval into an interval in the dual order. -/
+def dual : nonempty_interval α ≃ nonempty_interval αᵒᵈ :=
+{ to_fun := λ a, ⟨a.to_prod.swap, a.fst_le_snd⟩,
+  inv_fun := λ a, ⟨a.to_prod.swap, a.fst_le_snd⟩,
+  left_inv := λ a, ext _ _ $ prod.swap_swap _,
+  right_inv := λ a, ext _ _ $ prod.swap_swap _ }
+
+@[simp] lemma fst_dual (a : nonempty_interval α) : a.dual.fst = to_dual a.snd := rfl
+@[simp] lemma snd_dual (a : nonempty_interval α) : a.dual.snd = to_dual a.fst := rfl
+
+end has_le
+
+section preorder
+variables [preorder α] [preorder β] [preorder γ]
+
+instance : preorder (nonempty_interval α) := preorder.lift to_dual_prod
+
+/-- `{a}` as an interval. -/
+@[simps] def pure (a : α) : nonempty_interval α := ⟨⟨a, a⟩, le_rfl⟩
+
+lemma pure_injective : injective (pure : α → nonempty_interval α) :=
+λ a b, congr_arg $ prod.fst ∘ to_prod
+
+@[simp] lemma dual_pure (a : α) : (pure a).dual = pure (to_dual a) := rfl
+
+instance [inhabited α] : inhabited (nonempty_interval α) := ⟨pure default⟩
+instance : ∀ [nonempty α], nonempty (nonempty_interval α) := nonempty.map pure
+instance [nontrivial α] : nontrivial (nonempty_interval α) := pure_injective.nontrivial
+
+/-- Pushforward of nonempty intervals. -/
+@[simps] def map (f : α →o β) (a : nonempty_interval α) : nonempty_interval β :=
+⟨a.to_prod.map f f, f.mono a.fst_le_snd⟩
+
+@[simp] lemma map_pure (f : α →o β) (a : α) : (pure a).map f = pure (f a) := rfl
+@[simp] lemma map_map (g : β →o γ) (f : α →o β) (a : nonempty_interval α) :
+  (a.map f).map g = a.map (g.comp f) := rfl
+
+@[simp] lemma dual_map (f : α →o β) (a : nonempty_interval α) :
+  (a.map f).dual = a.dual.map f.dual := rfl
+
+variables [bounded_order α]
+
+instance : order_top (nonempty_interval α) :=
+{ top := ⟨⟨⊥, ⊤⟩, bot_le⟩,
+  le_top := λ a, ⟨bot_le, le_top⟩ }
+
+@[simp] lemma dual_top : (⊤ : nonempty_interval α).dual = ⊤ := rfl
+
+end preorder
+
+section partial_order
+variables [partial_order α]
+
+instance : partial_order (nonempty_interval α) := partial_order.lift _ to_dual_prod_injective
+
+/-- Consider a nonempty interval `[a, b]` as the set `[a, b]`. -/
+@[simps] def to_set : nonempty_interval α ↪o set α :=
+order_embedding.of_map_le_iff (λ a, Icc a.fst a.snd) (λ a b, Icc_subset_Icc_iff a.fst_le_snd)
```

Comment:
You could consider instantiating `set_like` on intervals. I can definitely imagine that people want to write `x ∈ [a, b]`.